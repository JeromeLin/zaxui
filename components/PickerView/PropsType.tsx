export interface BasePickerViewProps {
  visible?: boolean;
  dataSource?: any[];
  defaultValue?: string | any[];
  value?: string | any[];
  cols?: number;
  onChange?: (value?: any) => any;
  valueMember?: string;
  onValueChange?: (value?: any, i?: number) => void;
  itemRender?: (data?: object) => string;
}
