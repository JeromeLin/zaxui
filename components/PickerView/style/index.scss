@import "../../style/core/index";
@import "../../style/variables";

$prefixCls: $prefixCls-picker;
$picker-item-height: r(30);

.#{$prefixCls} {
  // width: 100%;
  // height: 100%;
  background: #fff;

  &-input {
    width: 100%;
    padding: r(12) 0;
  }

  &-placeholder {
    color: $placeholder-color;
  }

  &-disabled {
    color: $disabled-color;
  }

  &-wrapper {
    background: #fff;
  }

  &-header {
    display: flex;
    align-items: center;
    justify-content: center;
    position: relative;
    width: 100%;
    height: r(45);
    line-height: r(45);
    @include onepx(bottom);
  }

  &-cancel,
  &-submit {
    flex: 1;
    padding: 0 r(20);
    cursor: pointer;
  }

  &-cancel {
    color: #999;
  }

  &-submit {
    text-align: right;
    color: $theme-primary;
  }

  &-mask-top {
    -webkit-mask: -webkit-gradient(linear, 0% 40%, 0% 0%, from(rgba(222, 187, 71, 1)), to(rgba(36, 142, 36, 0)));
  }

  &-mask-bottom {
    -webkit-mask: -webkit-gradient(linear, 0% 60%, 0% 100%, from(rgba(222, 187, 71, 1)), to(rgba(36, 142, 36, 0)));
  }

  &-column-group {
    display: flex;
    align-items: center;

    &-item {
      flex: 1;
      text-align: center;
    }
  }

  &-column {
    display: block;
    position: relative;
    overflow: hidden;
    height: 7 * $picker-item-height;
    width: 100%;

    &-content {
      position: absolute;
      left: 0;
      top: 0;
      right: 0;
      padding: 3 * $picker-item-height 0;
    }

    &-item {
      text-align: center;
      font-size: r(16);
      line-height: $picker-item-height;
      height: $picker-item-height;
      color: #000;
      white-space: nowrap;
      text-overflow: ellipsis;
      overflow: hidden;

      &-selected {
        // font-size: r(18);
        // transition: all 0.3s;
      }
    }

    &-indicator {
      right: 0;
      position: absolute;
      left: 0;
      top: 3 * $picker-item-height;
      height: $picker-item-height;
      @include onepx(top bottom);
    }
  }

  &-crumbs {
    position: relative;
    padding: r(10) r(20);
    line-height: r(24);
    font-size: r(14);
    color: #999;
    @include onepx(bottom);

    p {
      margin: 0;
    }

    &-error {
      color: $theme-error;
    }
  }

}
