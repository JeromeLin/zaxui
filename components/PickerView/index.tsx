import React, { Component } from 'react';
import { formatBackToObject, initDataAndValue, updateDataSource, updateValue } from '../Picker/utils';
import Column from '../Column';
import { BasePickerViewProps } from './PropsType';

export interface PickerViewProps extends BasePickerViewProps {
  prefixCls?: string;
  className?: any;
}

export default class PickerView extends Component<PickerViewProps, any> {

  static defaultProps = {
    disabled: false,
    dataSource: [],
    onChange: () => {},
    prefixCls: 'za-picker',
    valueMember: 'value',
    itemRender: data => data.label,
  };

  constructor(props) {
    super(props);

    const initValue = props.value || props.defaultValue || [];

    let { data , value, cascade } = initDataAndValue(props.dataSource, initValue);

    this.state = {
      visible: props.visible || false,
      value,
      data,
      cascade,
    };
  }

  componentWillReceiveProps(nextProps) {
    if ('dataSource' in nextProps && nextProps.dataSource !== this.props.dataSource) {
      const { dataSource } = nextProps;
      let { data, cascade } = updateDataSource(dataSource);

      this.setState({
        data,
        cascade,
      });
    }

    if ('value' in nextProps && nextProps.value !== this.props.value) {
      const { dataSource, value } = nextProps;
      let { _value } = updateValue(dataSource, value);

      this.setState({
        value: _value,
      });
    }

  }

  onValueChange = (value) => {
    const { data, cascade } = this.state;
    const { onChange, onValueChange, valueMember, cols } = this.props;
    this.setState({
      value,
    });
    if (typeof onChange === 'function') {
      let _value: any;
      _value = formatBackToObject(data, value, cascade, valueMember, cols);
      onChange(_value);
    }

    if (typeof onValueChange === 'function') {
      onValueChange(value);
    }
  }

  render() {
    const { prefixCls, valueMember = PickerView.defaultProps.valueMember, itemRender } = this.props;
    const { data, value } = this.state;

    let PickerCol: JSX.Element;

    const cols = data.map((d) => {
      return { props: { children: d } };
    });

    if (this.state.cascade) {
      PickerCol = (
        <Column.Cascader
          prefixCls={prefixCls}
          data={data}
          value={value}
          cols={this.props.cols}
          itemRender={itemRender}
          valueMember={valueMember}
          onChange={v => this.onValueChange(v)}
        />
      );
    } else {
      PickerCol = (
        <Column.Group
          prefixCls={prefixCls}
          itemRender={itemRender}
          valueMember={valueMember}
          selectedValue={value}
          onValueChange={v => this.onValueChange(v)}
        >
          {cols}
        </Column.Group>
      );
    }

    return PickerCol;
  }
}
