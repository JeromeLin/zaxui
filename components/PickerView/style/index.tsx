import '../../style/';
import '../../Column/style';
import './index.scss';
