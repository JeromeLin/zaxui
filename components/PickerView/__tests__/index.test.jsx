import React from 'react';
import { render } from 'enzyme';
import toJson from 'enzyme-to-json';
import PickerView from '../index';

describe('PickerView', () => {
  it('PickerView render visible', () => {
    const wrapper = render(
      <PickerView
        dataSource={[
          { value: '1', label: '选项一' },
          { value: '2', label: '选项二' },
        ]}
        />
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
