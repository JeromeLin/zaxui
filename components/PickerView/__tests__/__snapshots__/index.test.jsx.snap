// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PickerView PickerView render visible 1`] = `
<div
  class="za-picker-mask-top"
>
  <div
    class="za-picker-mask-bottom"
  >
    <div
      class="za-picker-column-group"
    >
      <div
        class="za-picker-column-group-item"
      >
        <div
          class="za-picker-column"
        >
          <div
            class="za-picker-column-indicator"
          />
          <div
            class="za-picker-column-content"
          >
            <div
              class="za-picker-column-item za-picker-column-item-selected"
            >
              选项一
            </div>
            <div
              class="za-picker-column-item"
            >
              选项二
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
